### ISSUE TITLE
Order-dependent type inference failure with match

### ISSUE THREAD

**Issue by pythonesque** [rust-lang/rust#25165]
Not sure if this has been reported yet, but it really seems like Rust should be able to do better here.

```rust
trait Foo {
    fn foo(self) -> ();
}

impl Foo for () {
    fn foo(self) -> () { self }
}

fn main() {
    let mut res = None; // Compiles if we use `None::<()>`;
    match res {
        Some(winning) => { let _: () = winning.foo(); },
        None => res = Some(()), // Compiles if we move the None clause above the Some.
    }
}
```
---
```console
<anon>:12:40: 12:53 error: the type of this value must be known in this context
<anon>:12         Some(winning) => { let _: () = winning.foo(); },
```

**Comment by pythonesque** [comment]
I think https://github.com/rust-lang/rust/issues/22165 may be related too (in particular https://github.com/rust-lang/rust/issues/22165#issuecomment-75415175).

**Comment by pythonesque** [comment]
Looks identical to https://github.com/rust-lang/rust/issues/19855.  Would like clarification on what would be involved to fix this--I strongly disagree that this is expected behavior from the user side of things, even if it's expected with our current inference system.

**Comment by arielb1** [comment]
@pythonesque 

We would probably have to run inference in a loop of some sort.

**Comment by ebfull** [comment]
I ran into this in #29834. It definitely is a wart. In my case it prevents me from making wrappers around closures without breaking the inference logic:

```rust
struct Wrapper<F>(F);
struct Foo;

impl Foo {
    fn stuff(self) { }
}

fn api<F: Fn(Foo)>(f: Wrapper<F>) { }

fn main() {
    api(Wrapper(|b| {
        b.stuff();
    }))
}
```

... which makes my API really ugly. I too would like to know what's involved with changing this behavior. Perhaps the new MIR stuff can help?

**Comment by arielb1** [comment]
@ebfull 

MIR runs only after typechecking, so it can't help.

**Comment by arielb1** [comment]
cc @nikomatsakis

**Comment by nikomatsakis** [comment]
The reason that this doesn't work is that the type checker's current design includes various points where it cannot make progress unless it has enough type information. I think the fix would be to reimplement the type checker so that everything is a generic constraint that can be deferred; this way, we would just postpone processing `winning.foo()` until we learn the type of `winning` from the other arm. I've been wanting to do some general refactoring of typeck in this direction for some time, but it'd be a big project. I was hoping though to generally refactor typeck as part of the work on MIR -- specifically, to strip out region processing and move that to MIR -- and I was hoping to explore this at that time.

**Comment by arielb1** [comment]
@nikomatsakis 

This may be somewhat problematic because of coercions and autoderefs. Anyway, I feel like regionck is separated enough from typeck.

**Comment by nikomatsakis** [comment]
@arielb1 hmm, yes, I agree there are big challenges around coercion. I was hoping we could work around those by being capturing the source order somewhat in the graph, but it may be that it prevents the whole idea from really panning out.

I strongly disagree about regionck and typeck though. But we can discuss at another time, it's not really that relevant to this bug.

**Comment by pnkfelix** [comment]
(switched this from C-bug to C-enhancement, because I do not think this represent anything incorrect in our existing type inference; it just is one of many annoying short-comings that we might attempt to address in the future.)
